### PR TITLE
Deploy an example Node app to ECS

### PR DESCRIPTION
**Pull request by alex9smith**
Add a Cloud Formation template, SAM config and a Hello World Node.js app
that can be deployed to run on AWS ECS using Fargate.

This is the smallest template that's needed to run an app and access it
via a DNS record from the public internet. Not all of this is needed for
every deployed service however - the networking stack (VPC, subnets,
internet gateway, routing etc.) is all something that can be set up once
per account and used across multiple services. The aim of this example
however is to work out everything that's needed to run an internet
facing service, so it's all in a single template and stack.

Once it's deployed with `sam deploy`, the service can be accessed at the
DNS name for the load balancer.

Note, the Node app is present in this repo but currently there's no
automation or instructions for building the container and pushing it to
ECR. Instead I built the container on my laptop and ran `docker push` to
push it to my account on Docker Hub. This isn't sustainable as Docker
Hub rate limits downloads, but it's fine for an example app. We have
followup work planned to set up a proper pipeline.

---
[Trello](https://trello.com/c/GRDON4nX/1288-tech-spike-on-moving-account-service-dashboard-to-di-infrastructure)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+.env.development.local
+.env.test.local
+.env.production.local
+.env.local
+
+# parcel-bundler cache (https://parceljs.org/)
+.cache
+.parcel-cache
+
+# Next.js build output
```
Comment (huwd) [inline .gitignore:86]: Guessing this is boilerplate from somewhere?
Was this from AWS bootstrap or something we nicked off somewhere else?

Comment (alex9smith) [reply .gitignore:86]: Oh yeah - it's Github's boilerplate one https://github.com/github/gitignore/blob/main/Node.gitignore

Comment (alex9smith) [reply .gitignore:86]: They're CC0 so no problems with just using it

File: infrastructure/fargate-demo/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Fargate demo
+
+Deploy a Hello World Node app as a container on ECS / Fargate.
+
+## Getting AWS credentials
+
+At time of writing our new AWS account isn't in gds-cli yet so we have to do a bit of fiddling to get CLI credentials.
+
+```bash
+aws-vault exec gds-users -- aws sts assume-role --role-arn arn:aws:iam::626456592666:role/FIRST_NAME.LAST_NAME-admin --role-session-name test --region eu-west-2 --serial-number arn:aws:iam::622626885786:mfa/FIRST_NAME.LAST_NAME@digital.cabinet-office.gov.uk --token-code YOUR_MFA_CODE_HERE
```
Comment (huwd) [inline infrastructure/fargate-demo/README.md:10]: Will give this a go shortly - does the id `626456592666` vary too?

Comment (huwd) [reply infrastructure/fargate-demo/README.md:10]: and the serial too?

File: infrastructure/fargate-demo/template.yaml [hunk 3]
```diff
@@ -171,14 +171,18 @@ Resources:
...
+        - 0
+        - Fn::GetAZs: !Ref 'AWS::Region'
 
   Subnet2:
     Type: AWS::EC2::Subnet
     Properties:
       VpcId: !Ref VPC
       CidrBlock: 10.0.2.0/24
-      AvailabilityZone: "eu-west-2b"
+      AvailabilityZone: !Select
```
Comment (huwd) [inline infrastructure/fargate-demo/template.yaml:183]: Interesting, would be good to chat a bit more about this change?

Does this keep our app limited to any of the eu-west ones, or could it rock up on a us zone?
Wondering if we'll need to limit our stuff to particular zones, or if it's free to travel wherever there's availibility?

**Comment by huwd** [comment]
Looks great, congrats for getting that up.
I'll try and run the SAM command now and 👍🏻 the README.md

Would be good to chat more about that availbility zone change, check I understand the implications of what's going on there?

File: infrastructure/fargate-demo/README.md (hunk 2 above)
Comment (alex9smith) [reply infrastructure/fargate-demo/README.md:10]: Nope, those are the account IDs of our AWS accounts - the role one is our prototype account (where the role's defined) and the serial one is a reference to your IAM user in the gds-users AWS account.

File: infrastructure/fargate-demo/template.yaml (hunk 3 above)
Comment (alex9smith) [reply infrastructure/fargate-demo/template.yaml:183]: Ah good question! The `samconfig.toml` file defines the region to deploy to, so if you run `sam deploy` these will always end up in `eu-west-2` (London). But it seems best practice (ie. there's a linter warning) is to pick up availability zones and regions programatically so we could deploy to a different region if we ever wanted to

File: infrastructure/fargate-demo/README.md (hunk 2 above)
Comment (huwd) [reply infrastructure/fargate-demo/README.md:10]: OK ace I think I see what's going on here - looks like they're static

**Comment by alex9smith** [comment]
>  I'll try and run the SAM command now and 👍🏻 the README.md

You'll need to change the stack name in `samconfig.toml` if you want to deploy everything fresh, otherwise it'll try to overwrite and there won't be any changes (since this is already deployed)

File: infrastructure/fargate-demo/template.yaml (hunk 3 above)
Comment (huwd) [reply infrastructure/fargate-demo/template.yaml:183]: right, gotcha - that makes sense

**Review by huwd — APPROVED** [review]

**Pull request merged**